### PR TITLE
fix tiflash crash when setting join_concurrent_build = 0

### LINKED ISSUE #4734: tiflash crash when setting `join_concurrent_build` = 0
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

```

drop table if exists test.t1;
drop table if exists test.t2;
drop table if exists test.t1_not_null;
drop table if exists test.t2_not_null;
drop table if exists test.t1_decimal;
drop table if exists test.t2_decimal;

create table test.t1 (id int, value int);
create table test.t2 (id int, value int);
create table test.t1_not_null (id int not null, value int not null);
create table test.t2_not_null (id int not null, value int not null);
create table test.t1_decimal (id decimal(10,2) not null, value int not null);
create table test.t2_decimal (id decimal(6,3) not null, value int not null);

insert into test.t1 values(1,2),(3,4),(3,5),(3,6),(5,6),(7,8),(9,10),(null,11),(null,12),(10,null),(null,null);
insert into test.t2 values(3,4),(5,6),(5,7),(5,8),(null,7),(null,8);
insert into test.t1_not_null values(0,1),(1,2),(3,4),(4,5),(5,6);
insert into test.t2_not_null values(4,5),(5,6),(6,7);
insert into test.t1_decimal values(0,1),(1,2),(3,4),(4,5),(5,6);
insert into test.t2_decimal values(4,5),(5,6),(6,7);

alter table test.t1 set tiflash replica 1;
alter table test.t2 set tiflash replica 1;
alter table test.t1_not_null set tiflash replica 1;
alter table test.t2_not_null set tiflash replica 1;
alter table test.t1_decimal set tiflash replica 1;
alter table test.t2_decimal set tiflash replica 1;

analyze table test.t1;
analyze table test.t2;
analyze table test.t1_not_null;
analyze table test.t2_not_null;
analyze table test.t1_decimal;
analyze table test.t2_decimal;

use test; set session tidb_isolation_read_engines='tiflash'; set session tidb_opt_broadcast_join=1; select /*+ broadcast_join(t1,t2) */ t1.id,t1.value,t2.id,t2.value from t2 left join t1 on t2.id = t1.id;
```

### 2. What did you expect to see? (Required)
query success

### 3. What did you see instead (Required)
tiflash crash

Segmentation fault (core dumped)

```
[2022/04/22 11:27:06.290 +08:00] [ERROR] [BaseDaemon.cpp:423] ["BaseDaemon:Address not mapped to object."] [thread_id=37]
[2022/04/22 11:27:08.500 +08:00] [ERROR] [BaseDaemon.cpp:570] ["BaseDaemon:\n       0x1ebc3b1\tfaultSignalHandler(int, siginfo_t*, void*) [tiflash+32228273]\n                \tlibs/libdaemon/src/BaseDaemon.cpp:221\n  0x7f1d0695e630\t<unknown symbol> [libpthread.so.0+63024]\n  0x7f1d0a1c676d\t__dynamic_cast [libc++abi.so.1+243565]\n       0x7383a04\tDB::ExpressionBlockInputStream::getTotals() [tiflash+121125380]\n                \tdbms/src/DataStreams/ExpressionBlockInputStream.cpp:32\n       0x66724e8\tDB::IProfilingBlockInputStream::getTotals() [tiflash+107422952]\n                \tdbms/src/DataStreams/IProfilingBlockInputStream.cpp:419\n       0x66724e8\tDB::IProfilingBlockInputStream::getTotals() [tiflash+107422952]\n                \tdbms/src/DataStreams/IProfilingBlockInputStream.cpp:419\n       0x748cc73\tDB::CreatingSetsBlockInputStream::createOne(DB::SubqueryForSet&) [tiflash+122211443]\n                \tdbms/src/DataStreams/CreatingSetsBlockInputStream.cpp:258\n       0x1d9debc\tauto DB::wrapInvocable<std::__1::function<void ()> >(bool, std::__1::function<void ()>&&)::'lambda'()::operator()() [tiflash+31055548]\n                \tdbms/src/Common/wrapInvocable.h:36\n       0x1d9e025\tstd::__1::packaged_task<void ()>::operator()() [tiflash+31055909]\n                \t/usr/local/bin/../include/c++/v1/future:2089\n       0x1d8d5f8\tDB::DynamicThreadPool::executeTask(std::__1::unique_ptr<DB::IExecutableTask, std::__1::default_delete<DB::IExecutableTask> >&) [tiflash+30987768]\n                \tdbms/src/Common/DynamicThreadPool.cpp:101\n       0x1d8d4d6\tDB::DynamicThreadPool::dynamicWork(std::__1::unique_ptr<DB::IExecutableTask, std::__1::default_delete<DB::IExecutableTask> >) [tiflash+30987478]\n                \tdbms/src/Common/DynamicThreadPool.cpp:142\n       0x1d8f14d\tauto std::__1::thread DB::ThreadFactory::newThread<void (DB::DynamicThreadPool::*)(std::__1::unique_ptr<DB::IExecutableTask, std::__1::default_delete<DB::IExecutableTask> >), DB::DynamicThreadPool*, std::__1::unique_ptr<DB::IExecutableTask, std::__1::default_delete<DB::IExecutableTask> > >(bool, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, void (DB::DynamicThreadPool::*&&)(std::__1::unique_ptr<DB::IExecutableTask, std::__1::default_delete<DB::IExecutableTask> >), DB::DynamicThreadPool*&&, std::__1::unique_ptr<DB::IExecutableTask, std::__1::default_delete<DB::IExecutableTask> >&&)::'lambda'(auto&&...)::operator()<DB::DynamicThreadPool*, std::__1::unique_ptr<DB::IExecutableTask, std::__1::default_delete<DB::IExecutableTask> > >(auto&&...) const [tiflash+30994765]\n                \tdbms/src/Common/ThreadFactory.h:47\n       0x1d8ef29\tvoid* std::__1::__thread_proxy<std::__1::tuple<std::__1::unique_ptr<std::__1::__thread_struct, std::__1::default_delete<std::__1::__thread_struct> >, std::__1::thread DB::ThreadFactory::newThread<void (DB::DynamicThreadPool::*)(std::__1::unique_ptr<DB::IExecutableTask, std::__1::default_delete<DB::IExecutableTask> >), DB::DynamicThreadPool*, std::__1::unique_ptr<DB::IExecutableTask, std::__1::default_delete<DB::IExecutableTask> > >(bool, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, void (DB::DynamicThreadPool::*&&)(std::__1::unique_ptr<DB::IExecutableTask, std::__1::default_delete<DB::IExecutableTask> >), DB::DynamicThreadPool*&&, std::__1::unique_ptr<DB::IExecutableTask, std::__1::default_delete<DB::IExecutableTask> >&&)::'lambda'(auto&&...), DB::DynamicThreadPool*, std::__1::unique_ptr<DB::IExecutableTask, std::__1::default_delete<DB::IExecutableTask> > > >(void*) [tiflash+30994217]\n                \t/usr/local/bin/../include/c++/v1/thread:291\n  0x7f1d06956ea5\tstart_thread [libpthread.so.0+32421]"] [thread_id=37]
```

### 4. What is your TiFlash version? (Required)
master

### PR DESCRIPTION
**Pull request by SeaRise**
### What problem does this PR solve?

Issue Number: close #4734

Problem Summary:

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

File: dbms/src/Flash/Coprocessor/DAGQueryBlockInterpreter.cpp [hunk 1]
```diff
@@ -695,8 +695,11 @@ void DAGQueryBlockInterpreter::handleJoin(const tipb::Join & join, DAGPipeline &
 
     // add a HashJoinBuildBlockInputStream to build a shared hash table
     size_t stream_index = 0;
+    auto get_stream_index = [&stream_index, &join_build_concurrency]() {
+        return (stream_index++) % join_build_concurrency;
+    };
     right_pipeline.transform(
-        [&](auto & stream) { stream = std::make_shared<HashJoinBuildBlockInputStream>(stream, join_ptr, stream_index++, log->identifier()); });
+        [&](auto & stream) { stream = std::make_shared<HashJoinBuildBlockInputStream>(stream, join_ptr, get_stream_index(), log->identifier()); });
```
Comment (windtalker) [inline dbms/src/Flash/Coprocessor/DAGQueryBlockInterpreter.cpp:702]: The name `stream_index` is kind of indicating it should be unique between different `HashJoinBuildBlockInputStream`, I prefer to change the name or move the `stream_index%join_build_concurrency` to `Join.cpp`

Comment (SeaRise) [reply dbms/src/Flash/Coprocessor/DAGQueryBlockInterpreter.cpp:702]: But `join.insertFromBlock` will do `stream_index%join_build_concurrency` every time.
How about a different name? Like `join_build_pool_id`?

Comment (windtalker) [reply dbms/src/Flash/Coprocessor/DAGQueryBlockInterpreter.cpp:702]: LGTM

**Review by windtalker — APPROVED** [review]
LGTM

**Review by Lloyd-Pottiger — APPROVED** [review]
LGTM

**Pull request merged**